### PR TITLE
[Enhancement] Only create necessay partitions when refreshing mv by partition

### PR DESCRIPTION
**Pull request by nshangyiming**
This has been tested on 3.1 based branch in customer env, here is just a backport.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 30 / 31 (96.77%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|21|22|95.45%|[144]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|9|9|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 21 / 21 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|1|1|100.00%|[]|

**Review by murphyatwork — DISMISSED** [review]

**Review by LiShuMing — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java [hunk 1]
```diff
@@ -653,7 +653,7 @@ public static RangePartitionDiff getPartitionDiff(Expr partitionExpr, Column par
             if (functionCallExpr.getFnName().getFunction().equalsIgnoreCase(FunctionSet.DATE_TRUNC) ||
                     functionCallExpr.getFnName().getFunction().equalsIgnoreCase(FunctionSet.STR2DATE)) {
                 return SyncPartitionUtils.getRangePartitionDiffOfExpr(basePartitionMap,
-                        mvPartitionMap, functionCallExpr, partitionColumn.getPrimitiveType());
+                        mvPartitionMap, functionCallExpr, partitionColumn.getPrimitiveType(), null);
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java:656]: it seems that getRangePartitionDiffOfExpr without last null parameter is the same as with null. so it is not necessary to modify this line.

Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java:656]: The callee needs extra parameter, i don't want to add another method, so i will keep it unchanged for now.

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 21 / 21 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|1|1|100.00%|[]|

**Review by ABingHuang — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 21 / 21 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**